client/mm/libxc/binance.go

```diff
@@ -1079,209 +1233,163 @@ func (bnc *binance) getUserDataStream(ctx context.Context) (err error) {
+		return
 	}
-
-	return nil
+	book.updateQueue <- note.Data
 }
 
-func (bnc *binance) startMarketDataStream(ctx context.Context, baseSymbol, quoteSymbol string) (err error) {
-	slug := strings.ToLower(baseSymbol + quoteSymbol)
-
-	bnc.marketStreamMtx.Lock()
-	defer bnc.marketStreamMtx.Unlock()
-
-	// If a market stream already exists, just subscribe to this market.
-	if bnc.marketStream != nil {
-		bnc.booksMtx.Lock()
-		_, exists := bnc.books[slug]
-		if !exists {
-			bnc.books[slug] = newBNCBook()
-		} else {
-			bnc.books[slug].numSubscribers++
-		}
-		bnc.booksMtx.Unlock()
+func (bnc *binance) getOrderbookSnapshot(ctx context.Context, mktSymbol string) (*binanceOrderbookSnapshot, error) {
+	v := make(url.Values)
+	v.Add("symbol", strings.ToUpper(mktSymbol))
+	v.Add("limit", "1000")
+	resp := &binanceOrderbookSnapshot{}
+	return resp, bnc.getAPI(ctx, "/api/v3/depth", v, false, false, resp)
+}
 
-		if err := bnc.subUnsubDepth(bnc.marketStream, "SUBSCRIBE", slug); err != nil {
-			return fmt.Errorf("subUnsubDepth(SUBSCRIBE): %v", err)
-		}
-		return nil
+// subscribeToAdditionalMarketDataStream is called when a new market is
+// subscribed to after the market data stream connection has already been
+// established.
+func (bnc *binance) subscribeToAdditionalMarketDataStream(ctx context.Context, baseID, quoteID uint32) (err error) {
+	baseCfg, quoteCfg, err := bncAssetCfgs(baseID, quoteID)
+	if err != nil {
+		return fmt.Errorf("error getting asset cfg for %d: %w", baseID, err)
 	}
 
-	streamHandler := func(b []byte) {
-		var note *bncBookNote
-		if err := json.Unmarshal(b, &note); err != nil {
-			bnc.log.Errorf("Error unmarshaling book note: %v", err)
-			return
-		}
+	mktID := binanceMktID(baseCfg, quoteCfg)
+	streamID := marketDataStreamID(mktID)
 
-		if note == nil || note.Data == nil {
-			bnc.log.Debugf("No data in %q update: %+v", slug, note)
-			return
-		}
+	bnc.booksMtx.Lock()
+	defer bnc.booksMtx.Unlock()
 
-		parts := strings.Split(note.StreamName, "@")
-		if len(parts) != 2 || parts[1] != "depth20" {
-			bnc.log.Errorf("Unknown stream name %q", note.StreamName)
-			return
-		}
-		slug = parts[0] // will be lower-case
+	book, found := bnc.books[mktID]
+	if !found {
+		bnc.books[mktID] = newBinanceOrderBook(baseCfg.conversionFactor, quoteCfg.conversionFactor, mktID, bnc.log)
+		book = bnc.books[mktID]
+	} else {
+		book.mtx.Lock()
+		book.numSubscribers++
+		book.mtx.Unlock()
+		return
+	}
 
-		bnc.log.Tracef("Received %d bids and %d asks in a book update for %s", len(note.Data.Bids), len(note.Data.Asks), slug)
+	if err := bnc.subUnsubDepth(true, streamID); err != nil {
+		return fmt.Errorf("error subscribing to %s: %v", streamID, err)
+	}
 
-		bids, err := bncParseBookUpdates(note.Data.Bids)
-		if err != nil {
-			bnc.log.Errorf("Error parsing bid updates: %v", err)
-			return
-		}
+	getSnapshot := func() (*binanceOrderbookSnapshot, error) {
+		return bnc.getOrderbookSnapshot(ctx, mktID)
+	}
+	go book.sync(ctx, getSnapshot)
```

Comment:
What do you mean by watch the routine?

Reply:
I mean the `go book.sync(ctx, getSnapshot)` goroutine. The program will not wait for it to finish on shutdown, probably. It could also be viewed as code bloat to add a waitgroup, so feel free to ignore. Maybe there's no benefit to waiting at all. If it were writing to a file maybe? I don't think it is though.